src/tqec/sketchup/zx_graph.py

```diff
@@ -0,0 +1,334 @@
+        """Return a list of nodes in the graph."""
+        return [data[_NODE_DATA_KEY] for _, data in self._graph.nodes(data=True)]
+
+    @property
+    def edges(self) -> list[ZXEdge]:
+        """Return a list of edges in the graph."""
+        return [data[_EDGE_DATA_KEY] for _, _, data in self._graph.edges(data=True)]
+
+    def add_node(
+        self,
+        position: Position3D,
+        node_type: NodeType,
+    ) -> None:
+        """Add a node to the graph.
+
+        Args:
+            position: The 3D position of the node.
+            node_type: The type of the node.
+        """
+        if position in self._graph:
+            raise TQECException(f"Node {position} already exists in the graph.")
+        self._graph.add_node(position, **{_NODE_DATA_KEY: ZXNode(position, node_type)})
+
+    def add_z_node(self, position: Position3D) -> None:
+        """Add a Z-type node to the graph."""
+        self.add_node(position, NodeType.Z)
+
+    def add_x_node(self, position: Position3D) -> None:
+        """Add an X-type node to the graph."""
+        self.add_node(position, NodeType.X)
+
+    def add_virtual_node(self, position: Position3D) -> None:
+        """Add a virtual node to the graph."""
+        self.add_node(position, NodeType.V)
+
+    def add_edge(
+        self,
+        u: Position3D,
+        v: Position3D,
+        has_hadamard: bool = False,
+    ) -> None:
+        """Add an edge to the graph.
+
+        Args:
+            u: The position of the first node.
+            v: The position of the second node.
+            has_hadamard: Whether the edge has a Hadamard transition.
+        """
+        if u not in self._graph or v not in self._graph:
+            raise TQECException("Both nodes must exist in the graph.")
+        u_node = self._graph.nodes[u][_NODE_DATA_KEY]
+        v_node = self._graph.nodes[v][_NODE_DATA_KEY]
+        self._graph.add_edge(
+            u, v, **{_EDGE_DATA_KEY: ZXEdge(u_node, v_node, has_hadamard)}
+        )
+
+    def get_node(self, position: Position3D) -> ZXNode | None:
+        """Get the node by position."""
+        if position not in self._graph:
+            return None
+        return self._graph.nodes[position][_NODE_DATA_KEY]
+
+    def get_edge(self, u: Position3D, v: Position3D) -> ZXEdge | None:
+        """Get the edge by its endpoint positions."""
+        if not self._graph.has_edge(u, v):
+            return None
+        return self._graph.edges[u, v][_EDGE_DATA_KEY]
+
+    def edges_at(self, position: Position3D) -> list[ZXEdge]:
+        """Get the edges incident to a node."""
+        return [
+            data[_EDGE_DATA_KEY]
+            for _, _, data in self._graph.edges(position, data=True)
+        ]
+
+    def draw(self, show_title: bool = True) -> None:
+        """Draw the 3D graph using matplotlib."""
+        import matplotlib.pyplot as plt
+
+        fig = plt.figure()
+        ax = fig.add_subplot(111, projection="3d")
```

Comment:
Fixing a few typing errors returned by ruff (not mypy yet)
```suggestion
    def draw(self, show_title: bool = True) -> None:
        """Draw the 3D graph using matplotlib."""
        import matplotlib.pyplot as plt
        from mpl_toolkits.mplot3d.axes3d import Axes3D

        fig = plt.figure()
        # See https://matplotlib.org/stable/users/explain/toolkits/mplot3d.html
        ax = typing.cast(Axes3D, fig.add_subplot(111, projection="3d"))
```

Reply:
Disregard that, I will do a proper review when you think the PR is ready :)